crypto/threads_none.c

```diff
@@ -0,0 +1,138 @@
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * 3. All advertising materials mentioning features or use of this
+ *    software must display the following acknowledgment:
+ *    "This product includes software developed by the OpenSSL Project
+ *    for use in the OpenSSL Toolkit. (http://www.openssl.org/)"
+ *
+ * 4. The names "OpenSSL Toolkit" and "OpenSSL Project" must not be used to
+ *    endorse or promote products derived from this software without
+ *    prior written permission. For written permission, please contact
+ *    openssl-core@openssl.org.
+ *
+ * 5. Products derived from this software may not be called "OpenSSL"
+ *    nor may "OpenSSL" appear in their names without prior written
+ *    permission of the OpenSSL Project.
+ *
+ * 6. Redistributions of any form whatsoever must retain the following
+ *    acknowledgment:
+ *    "This product includes software developed by the OpenSSL Project
+ *    for use in the OpenSSL Toolkit (http://www.openssl.org/)"
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE OpenSSL PROJECT ``AS IS'' AND ANY
+ * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
+ * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ====================================================================
+ */
+
+#include <openssl/crypto.h>
+
+#if !defined(OPENSSL_THREADS) || defined(CRYPTO_TDEBUG)
+
+int CRYPTO_MUTEX_init(CRYPTO_MUTEX *lock)
+{
+    int *l = lock;
+    *l = 1;
+
+    return 1;
+}
+
+int CRYPTO_MUTEX_lock_read(CRYPTO_MUTEX *lock)
+{
+    OPENSSL_assert(*(int *) lock == 1);
+    return 1;
+}
+
+int CRYPTO_MUTEX_lock_write(CRYPTO_MUTEX *lock)
+{
+    OPENSSL_assert(*(int *) lock == 1);
+    return 1;
+}
+
+int CRYPTO_MUTEX_unlock(CRYPTO_MUTEX *lock)
+{
+    OPENSSL_assert(*(int *) lock == 1);
+    return 1;
+}
+
+void CRYPTO_MUTEX_cleanup(CRYPTO_MUTEX *lock) { }
+
+void CRYPTO_ONCE_run(CRYPTO_ONCE *once, void (*init)(void))
+{
+    if (*once)
+        return;
+
+    init();
+    *once = 1;
+}
+
+#define OPENSSL_CRYPTO_THREAD_LOCAL_KEY_MAX 256
+
+static void *thread_local_storage[OPENSSL_CRYPTO_THREAD_LOCAL_KEY_MAX];
+static void *thread_local_cleanup[OPENSSL_CRYPTO_THREAD_LOCAL_KEY_MAX];
```

Comment:
Array of void pointers is cleaner from a purely cosmetic perspective, but .you are probably right. It's not used right now, so it could maybe be removed completely.

I originally added it to implement a CRYPTO_THREAD_LOCAL_cleanup function, but in the end I didn't do it. Although, since no-threads/Windows don't deallocate local storage it's probably worth looking into.